lib/rules/no-side-effects-wait-for.ts

```diff
@@ -0,0 +1,65 @@
+import { ESLintUtils, TSESTree } from '@typescript-eslint/experimental-utils'
+import { getDocsUrl } from '../utils'
+import { isBlockStatement, findClosestCallNode, isMemberExpression, isCallExpression, isIdentifier } from '../node-utils'
+
+export const RULE_NAME: string = 'no-side-effects-wait-for';
+
+const WAIT_EXPRESSION_QUERY: string =
+  'CallExpression[callee.name=/^(waitFor)$/]';
+
+const SIDE_EFFECTS: Array<string> = ['fireEvent', 'userEvent']
+
+export type MessageIds = 'noSideEffectsWaitFor';
+type Options = [];
+
+export default ESLintUtils.RuleCreator(getDocsUrl)<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        "It's preferred to avoid side effects in `waitFor`",
+      category: 'Best Practices',
+      recommended: false,
+    },
+    messages: {
+      noSideEffectsWaitFor: 'Avoid using side effects within `waitFor` callback',
+    },
+    fixable: null,
+    schema: [],
+  },
+  defaultOptions: [],
+  create: function(context) {
+    function reportSideEffects(
+      node: TSESTree.BlockStatement
+    ) {
+      const totalSideEffects = (body: Array<TSESTree.Node>): Array<TSESTree.Node> =>
+        body.filter((node: TSESTree.ExpressionStatement) => {
```

Comment:
we could probably use `some()` here, which already returns a boolean, and remove the check for `.length > 0` in line 51. Because of that, the method name could be changed to something like `hasSideEffects` or similar